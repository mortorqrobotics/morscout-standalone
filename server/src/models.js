import { model } from "mongoose";
import allianceSchema from "shared/schemas/Alliance";
import matchSchema from "shared/scheams/Match";
import regionalSchema from "shared/schemas/Regional";
import seasonSchema from "shared/scheams/Season";
import teamSchema from "shared/schemas/Team";
import userSchema from "shared/schemas/User";

export const Alliance = model("Alliance", allianceSchema);
export const Match = model("Match", matchSchema);
export const Regional = model("Regional", regionalSchema);
export const Season = model("Season", seasonSchema);
export const Team = model("Team", teamSchema);
export const User = model("User", userSchema);
