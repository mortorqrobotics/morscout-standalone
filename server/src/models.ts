import { model, Model } from "mongoose";
import allianceSchema, { IAlliance } from "Shared/schemas/Alliance";
import matchSchema, { IMatch } from "Shared/schemas/Match";
import regionalSchema, { IRegional } from "shared/schemas/Regional";
import seasonSchema, { ISeason } from "Shared/schemas/Season";
import teamSchema, { ITeam } from "Shared/schemas/Team";
import userSchema, { IUser } from "Shared/schemas/User";

export const Alliance: Model<IAlliance> = model("Alliance", allianceSchema);
export const Match: Model<IMatch> = model("Match", matchSchema);
export const Regional: Model<IRegional> = model("Regional", regionalSchema);
export const Season: Model<ISeason> = model("Season", seasonSchema);
export const Team: Model<ITeam> = model("Team", teamSchema);
export const User: Model<IUser> = model("User", userSchema);
