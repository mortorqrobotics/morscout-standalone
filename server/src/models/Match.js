import { Schema } from "mongoose";

const Match = new Schema({
  redAlliance: {
    type: Schema.Types.ObjectId,
    ref: "Alliance"
  },
  blueAlliance: {
    type: Schema.Types.ObjectId,
    ref: "Alliance"
  },
  winner: {
    type: String,
    enum: ["Red", "Blue"],
    default: null
  },
  startDate: {
    type: Date
  },
  endData: {
    type: Date,
    validate: date => this.startDate.time() < date.time()
  },
  Regional: {
    type: Schema.Types.ObjectId,
    ref: "Regional"
  }
});

export default Match;
