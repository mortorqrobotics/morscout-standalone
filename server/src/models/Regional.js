import { Schema } from "mongoose";
import { pointSchema, polygonSchema } from "./util/GeoJSON";
import { eventTypes } from "./util/TBA";

export default mongoose => {
  const RegionalSchema = new Schema({
    name: String,
    location: pointSchema,
    district: polygonSchema,
    url: {
      type: String,
      validate: url => encodeURI(url) === url
    },
    eventType: {
      type: Number,
      validate: type => Object.keys(eventTypes).includes(type)
    },
    startDate: {
      type: Date
    },
    endData: {
      type: Date,
      validate: date => this.startDate.time() < date.time()
    },
    season: {
      type: Schema.types.ObjectId,
      ref: "Season"
    }
  });

  return mongoose.model("Regional", RegionalSchema);
};
