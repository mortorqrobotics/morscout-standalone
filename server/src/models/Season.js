import { Schema } from "mongoose";

export default mongoose => {
  const SeasonSchema = new Schema({
    year: {
      type: Number,
      validate: year => year >= 1989 && year <= Date.now().getUTCFullYear() + 1,
      default: Date.now
    },
    name: String,
    buildStart: {
      type: Date,
      validate: date =>
        date.getUTCFullYear() === this.year ||
        date.getUTCFullYear() === this.year - 1
    },
    buildEnd: {
      type: Date,
      validate: date =>
        (date.getUTCFullYear() === this.year ||
          date.getUTCFullYear() === this.year - 1) &&
        date.time() > this.buildStart.time()
    },
    seasonStart: {
      type: Date,
      validate: date =>
        date.getUTCFullYear() === this.year &&
        date.time() > this.buildEnd.time()
    },
    seasonEnd: {
      type: Date,
      validate: date => date.time() > this.seasonStart.time()
    }
  });

  return mongoose.model("Season", SeasonSchema);
};
