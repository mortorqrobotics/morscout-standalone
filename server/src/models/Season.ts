import * as mongoose from "mongoose";

const Season: mongoose.Schema = new mongoose.Schema({
  year: {
    type: Number,
    validate: year => year >= 1989 && year <= (new Date()).getUTCFullYear() + 1,
    default: () => (new Date()).getUTCFullYear()
  },
  name: String,
  buildStart: {
    type: Date,
    validate: date =>
      date.getUTCFullYear() === this.year ||
      date.getUTCFullYear() === this.year - 1
  },
  buildEnd: {
    type: Date,
    validate: date =>
      (date.getUTCFullYear() === this.year ||
        date.getUTCFullYear() === this.year - 1) &&
      date.time() > this.buildStart.time()
  },
  seasonStart: {
    type: Date,
    validate: date =>
      date.getUTCFullYear() === this.year && date.time() > this.buildEnd.time()
  },
  seasonEnd: {
    type: Date,
    validate: date => date.time() > this.seasonStart.time()
  }
});

export default mongoose.model("Season", Season);
