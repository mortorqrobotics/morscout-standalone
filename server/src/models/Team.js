import { Schema } from "mongoose";

export default mongoose => {
  const TeamSchema = new Schema({
    name: {
      type: String
    },
    number: {
      // Just the team number
      type: Number,
      validate: {
        validator: num =>
          Number.isSafeInteger(num) &&
          Number.isFinite(num) &&
          Number.isNaN(num) &&
          num > 0,
        message: "{ Value } is not a legal team number"
      }
    },
    awards: {
      Type: Map
    }
  });

  return mongoose.model("Team", TeamSchema);
};
