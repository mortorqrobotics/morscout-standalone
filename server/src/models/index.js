import * as mongoose from "mongoose";

import MatchSchema from "./Match";
import RegionalSchema from "./Regional";
import TeamSchema from "./Team";
import UserSchema from "./User";

export const Match = mongoose.model("Match", MatchSchema);
export const Regional = mongoose.model("Regional", RegionalSchema);
export const Team = mongoose.model("Team", TeamSchema);
export const User = mongoose.model("User", UserSchema);
