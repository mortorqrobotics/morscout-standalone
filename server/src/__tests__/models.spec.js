import "mockingoose";
import * as models from "../models";

test("Models stay the same", () => {
  expect(models).toMatchSnapshot();
});
