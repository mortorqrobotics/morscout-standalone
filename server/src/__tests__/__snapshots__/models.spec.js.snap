// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Models stay the same 1`] = `
Object {
  "Alliance": [Function],
  "Match": [Function],
  "Regional": [Function],
  "Season": [Function],
  "Team": [Function],
  "User": [Function],
}
`;
